File: examples/aoc/2025/1/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func mod(x, y int64) int64 {
	return ((x % y + y) % y)
}

func parse(line string) int64 {
	if line == "" {
		return 0
	}
	val, _ := strconv.ParseInt(line[1:], 10, 64)
	if line[0] == 'L' {
		return -val
	}
	return val
}

func part1(nums []int64) {
	dial := int64(50)
	count := int64(0)
	for _, x := range nums {
		dial = mod(dial+x, 100)
		if dial == 0 {
			count++
		}
	}
	fmt.Printf("part_1: %d\n", count)
}

func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

func part2(nums []int64) {
	dial := int64(50)
	count := int64(0)
	for _, x := range nums {
		if dial > 0 && x < 0 && dial+x < 0 {
			count++
		}
		adj := abs(dial+x) - 1
		count += adj / 100
		dial = mod(dial+x, 100)
		if dial == 0 {
			count++
		}
	}
	fmt.Printf("part_2: %d\n", count)
}

func main() {
	data, _ := os.ReadFile("input.txt")
	parts := strings.Split(string(data), "\n")
	nums := make([]int64, len(parts))
	for i, line := range parts {
		nums[i] = parse(line)
	}
	part1(nums)
	part2(nums)
}

File: examples/aoc/2025/1/main.rs
fn mod_op(x: i64, y: i64) -> i64 {
    ((x % y) + y) % y
}

fn parse(line: &str) -> i64 {
    if line.is_empty() {
        return 0;
    }
    let val: i64 = line[1..].parse().unwrap();
    if line.as_bytes()[0] == b'L' { -val } else { val }
}

fn part_1(nums: &[i64]) {
    let mut dial = 50i64;
    let mut count = 0i64;
    for &x in nums {
        dial = mod_op(dial + x, 100);
        if dial == 0 { count += 1; }
    }
    println!("part_1: {count}");
}

fn part_2(nums: &[i64]) {
    let mut dial = 50i64;
    let mut count = 0i64;
    for &x in nums {
        if dial > 0 && x < 0 && dial + x < 0 { count += 1; }
        let adj = (dial + x).abs() - 1;
        count += adj / 100;
        dial = mod_op(dial + x, 100);
        if dial == 0 { count += 1; }
    }
    println!("part_2: {count}");
}

fn main() {
    let content = std::fs::read_to_string("input.txt").unwrap();
    let numbers: Vec<i64> = content.split('\n').map(parse).collect();
    part_1(&numbers);
    part_2(&numbers);
}
